### PR TITLE
if exit with EX_OK, won't mark it as worker lost, it's just normal (w…

### PR DESCRIPTION
**Pull request by tinylambda**
https://github.com/celery/celery/issues/3631

### CODE REVIEW

**Comment by tinylambda** [comment]
Why there are so many errors in the CI progress? anyone can help ?

**Comment by tinylambda** [comment]
@auvipy can you review this PR ?

**Comment by georgepsarakis** [comment]
@tinylambda the failures seem to be caused by https://github.com/jaraco/configparser/issues/30

I will restart the build jobs.

**Comment by tinylambda** [comment]
> @tinylambda the failures seem to be caused by [jaraco/configparser#30](https://github.com/jaraco/configparser/issues/30)
> 
> I will restart the build jobs.

Thanks very

**Comment by tinylambda** [comment]
> @auvipy can you review this PR ?

@ask @thedrow

File: celery/concurrency/asynpool.py [hunk 1]
```diff
@@ -1085,6 +1086,11 @@ def on_job_process_lost(self, job, pid, exitcode):
         was assigned to exited by mysterious means (error exitcodes and
         signals).
         """
+        # if exit with EX_OK, won't mark it as worker lost,
```
Comment (georgepsarakis) [inline celery/concurrency/asynpool.py]: If I am not mistaken, at this point it is not clear whether the process exited **after the job was completed**. Perhaps you also need to check if the `job.ready` [method](https://github.com/celery/billiard/blob/13a52dc1c3dd9ba2f01c255064b33ceb71a9cb85/billiard/pool.py#L1136) returns `True`.

Comment (tinylambda) [reply celery/concurrency/asynpool.py]: Yes, it's really important to make sure the job is completes before the process exits (which is the warm shutdown case) !

I will improve it, thanks ！

Comment (tinylambda) [reply celery/concurrency/asynpool.py]: I tested again and found that job.ready() is False, but exitcode is EX_OK, is it a expected state ?

I think the job process exit with EX_OK indicates that the job it executes is also OK, any suggestions ?

> If I am not mistaken, at this point it is not clear whether the process exited **after the job was completed**. Perhaps you also need to check if the `job.ready` [method](https://github.com/celery/billiard/blob/13a52dc1c3dd9ba2f01c255064b33ceb71a9cb85/billiard/pool.py#L1136) returns `True`.

Comment (georgepsarakis) [reply celery/concurrency/asynpool.py]: @tinylambda I am quite certain that it is not necessary. A process can shutdown gracefully without necessarily completing the task.

Comment (thedrow) [reply celery/concurrency/asynpool.py]: Please rephrase the comment.
"If the process exists with the EX_OK status code, don't mark the worker as lost"
I think it's clearer that way.
Also,for future reference please mention the relevant issue number.

File: celery/concurrency/asynpool.py [hunk 2]
```diff
@@ -1085,6 +1086,11 @@ def on_job_process_lost(self, job, pid, exitcode):
...
+        # it's just normal (warm) shutdown
+        if exitcode == EX_OK and job.ready():
```
Comment (thedrow) [inline celery/concurrency/asynpool.py]: Instead of the comment, maybe we should use a helper method?
It would be clearer.

**Comment by thedrow** [comment]
@tinylambda Are you going to finish this PR?

**Comment by tinylambda** [comment]
> @tinylambda Are you going to finish this PR?

Hi @thedrow ， Sorry for the later response.  I think there should be a much better solution to this issue, but I think, it needs a lot of knowledge on the design of Celery architecture and its implementations.

So now, I don't think my PR on this issue is a proper solution. 

I think it's OK to close this PR, and I will supply a better solution on this issue if I have enough time.

**Pull request closed without merging**